feat: `--<lang>_prefix` cli option

igor-sirotin opened the pull request:
Closes https://github.com/protocolbuffers/protobuf/issues/23509

# Description

- Add per-plugin `--<lang>_prefix=COMMAND` flag to `protoc`, allowing a plugin to be invoked through a user-supplied command wrapper.
- `<lang>` matches the plugin used by `--<lang>_out`, so different plugins can use different wrappers in a single invocation.
- protoc executes `COMMAND <plugin>`; COMMAND's first token is resolved via the search path (PATH), the value is tokenized on whitespace into argv, and quotes are rejected at parse time. Use a wrapper script for more complex invocations.
- Wrapper still honours custom plugin paths set with `--plugin`.
- Cross-platform tests (POSIX shell scripts; Windows `.bat`) cover happy-path invocation, argument forwarding, PATH-based resolution, missing executable, duplicate flags, empty values, and quote rejection.

# Naming

Earlier revisions of this PR used a single global `--plugin-command-prefix` flag. Per https://github.com/protocolbuffers/protobuf/issues/23509#issuecomment-3478221497, this PR now uses per-plugin `--<lang>_prefix` to match the existing `--<lang>_out` / `--<lang>_opt` convention.

# Testing

I've tested this in https://github.com/status-im/status-go with this command in `messaging/layers/encryption/`:
```
protoc --go_prefix="go tool" --go_out=. ./messaging/layers/encryption/protocol_message.proto
```

igor-sirotin committed:
feat: --plugin-command-prefix

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface.cc:
```diff
@@ -2453,6 +2453,25 @@ CommandLineInterface::InterpretArgument(const std::string& name,
     fatal_warnings_ = true;
+  } else if (name == "--plugin-command-prefix") {
+    if (plugin_prefix_.empty()) {
+      std::cerr << "This compiler does not support plugins." << std::endl;
```
nit: this error message is a little misleading.  It's really just that the user didn't specify a plugin

igor-sirotin replied:
I think the error is actually accurate. `plugin_prefix_` is an internal variable, which is defined by `AllowPlugins` method: https://github.com/protocolbuffers/protobuf/blob/3c164c1fed28a987f01fb387860348f5840a1aed/src/google/protobuf/compiler/command_line_interface.h#L179

This error and behaviour is copied from here: 
https://github.com/protocolbuffers/protobuf/blob/3c164c1fed28a987f01fb387860348f5840a1aed/src/google/protobuf/compiler/command_line_interface.cc#L2475-L2479

mkruskal-google replied:
Ah yes my bad, I misunderstood this.  tbh I wasn't even aware this plugin_prefix thing existed :)

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface.cc:
```diff
@@ -2453,6 +2453,25 @@ CommandLineInterface::InterpretArgument(const std::string& name,
+      std::cerr << name << " may only be passed once." << std::endl;
+      return PARSE_ARGUMENT_FAIL;
+    }
+    if (value.empty()) {
```
This is probably redundant with the check below

igor-sirotin replied:
Indeed. Removed in 82385939ec367ca150391c85c3be29e9e2e0d389.

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface.cc:
```diff
@@ -3065,10 +3088,24 @@ bool CommandLineInterface::GeneratePluginOutput(
+    std::vector<std::string> args(plugin_command_prefix_.begin() + 1,
+                                  plugin_command_prefix_.end());
+    args.push_back(executable);
+    subprocess.Start(plugin_command_prefix_.front(), Subprocess::SEARCH_PATH,
```
Would this be simpler if we just required EXACT_NAME for this case?  It would be a little more awkward on the CLI but it would be much clearer.  I wasn't even aware this SEARCH_PATH path existed and I'm not sure off the top of my head what it uses

igor-sirotin replied:
With `SEARCH_PATH` it looks up in `PATH`: https://github.com/protocolbuffers/protobuf/blob/3c164c1fed28a987f01fb387860348f5840a1aed/src/google/protobuf/compiler/subprocess.h#L45-L46

If we don't use `SEARCH_PATH`, then users would have to specify absolute paths, and I'm afraid it won't be quite convenient.

E.g. in case with Go, we want `--plugin-command-prefix "go tool"`. Forcing absolute path to `go` would be a problem. Especially if this command is git-tracked, like when using `protoc` [like this:](https://github.com/status-im/status-go/blob/d6da1e791841df36bfb6d942a2129e1ad7357d54/pkg/messaging/layers/encryption/protocol.go#L24)
```go
//go:generate protoc --plugin-command-prefix "go tool" --go_out=. ./foo.proto
```

mkruskal-google replied:
hmm I did an experiment and it looks like we have *zero* test coverage for this path.  If I just add an assertion (https://github.com/protocolbuffers/protobuf/pull/27193) where we run a plugin from the path *and* in Subprocess::Start, it never gets hit in any of our tests >.<. It's a little amazing we've never broken this, but we can follow up on this.

I agree this seems like a slightly nicer default behavior, although you could just do something like `$(which go) tool` right?

igor-sirotin replied:
Indeed `$(which go)` will work when running from shell. 
But in some peculiar cases it won't work 🙁 

E.g. in Go, `protoc` commands are often combined with `//go:generate` instructions:
```
//go:generate protoc --go_prefix "$(which go) tool" --go_out=. ./protocol_message.proto
```

I'ved tried it out, and using `$(which go)` is not expanded in such case, leading to expected error:
```
$(which: program not found or is not executable
Please specify a program using absolute path or make sure the program is available in your PATH system variable
```

Although even in this case it can be work-arounded using a wrapper script, I think it's better to keep `SEARCH_PATH`. It's not a big deal on code side, while it's a nicer default behaviour indeed.

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface_unittest.cc:
```diff
@@ -576,6 +576,39 @@ TEST_F(CommandLineInterfaceTest, GeneratorAndPlugin) {
+  ASSERT_EQ(0, chmod(wrapper_path.c_str(), 0777));
+#endif
+
+  Run(absl::StrCat("protocol_compiler --plugin-command-prefix=", wrapper_path,
```
It might be worth adding a few more tests or making this one more complicated.  Things to cover:

- wrappers with arguments
- wrappers looked up from the search path
- error conditions

igor-sirotin replied:
Added in a0bf6a59992fe012dc8674126f9f97515be1f7cc

mkruskal-google replied:
Nice :)

mkruskal-google commented on src/google/protobuf/compiler/subprocess.cc:
```diff
@@ -95,15 +99,56 @@ void Subprocess::Start(const std::string& program, SearchMode search_mode) {
   }
 
+  // CreateProcessW takes a single command-line string, so we need to quote any
+  // argument that contains spaces, tabs, quotes, or backslashes before quotes.
```
Can we just require users to quote these themselves?  This is a lot of complex untested code

igor-sirotin replied:
This actually seem to be a deeper problem. I'm thinking of a simple solution.
But we can certainly require users to do more think to simplify the code 👍

igor-sirotin replied:
I've removed this complexity in ef332a9b16f6837bbe9eb022bd02e66c3e17ce08.

Spaces in `--plugin-command-prefix` are now simply not supported, this simplifies the code a lot. When needed, users can introduce script wrappers.
https://github.com/protocolbuffers/protobuf/blob/ef332a9b16f6837bbe9eb022bd02e66c3e17ce08/src/google/protobuf/compiler/command_line_interface.cc#L2728-L2735

mkruskal-google replied:
Yea that seems like a much simpler first pass.  We could revisit that in the future, but this seems like it would be sufficient for >90% of the use-cases

mkruskal-google reviewed (commented):
Sorry for the long delay on this!

igor-sirotin committed:
Drop redundant empty-value check in --plugin-command-prefix

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface.cc:
```diff
@@ -2709,7 +2728,11 @@ Parse PROTO_FILES and generate output based on the options given:
                               is mapped to the given executable even if
-                              the executable's own name differs.)";
+                              the executable's own name differs.
+  --plugin-command-prefix=COMMAND
```
Two points on this naming:

1) plugin-prefix is a better name, and I think we can just rename the internal variable it collides with or give it a mismatched variable name

2) this has confusing semantics when multiple plugins are specified.  What if only one of them needs a wrapper?

I think an alternative here would be to have protoc take something like `<lang>_prefix`, and only use that prefix for `<lang>_out` generation.  It would require an extra mapping, but it would avoid the issues in 2)

igor-sirotin replied:
Working on this.

igor-sirotin replied:
`--<lang>_prefix` is a great idea, implemented it in 35e4a467d8bdd9219ef6b47506fbc26428e03dea. I'm envy I didn't come up with it 😄

I'm only unsure what we should do with help text. Printing all possible options (`--go_prefix`, `--cpp_prefix`, etc) sounds unnecessary. Do you think a generic help for `--<lang>_prefix` is ok?
https://github.com/protocolbuffers/protobuf/blob/35e4a467d8bdd9219ef6b47506fbc26428e03dea/src/google/protobuf/compiler/command_line_interface.cc#L2729-L2731

mkruskal-google replied:
Hmm good question.  I like your solution.  I was thinking we could fold this into the loop right below, but we actually don't *know* all the available plugins when people run `--help`.  All we know is the built-in generators, which this doesn't even work on.  It's also noisy as you pointed out

mkruskal-google replied:
kudos to @haberman for flagging the issue with 2

igor-sirotin committed:
drop bespoke Windows arg-quoting; preserve --plugin Win32 behavior

igor-sirotin committed:
Replace --plugin-command-prefix with per-plugin --<lang>_prefix

igor-sirotin committed:
Expand --<lang>_prefix test coverage

igor-sirotin commented:
@mkruskal-google Thank you for the comments, I've addressed all of them. 
Please take another look when you have a moment

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface.cc:
```diff
@@ -2709,7 +2725,14 @@ Parse PROTO_FILES and generate output based on the options given:
-                              the executable's own name differs.)";
+                              the executable's own name differs.
+  --<lang>_prefix=COMMAND     Runs the plugin used by --<lang>_out via
+                              COMMAND. protoc executes "COMMAND <plugin>"
```
We should probably mention that the COMMAND uses the search path

igor-sirotin replied:
Added in 6c9600026cd9046105747d48ec9ba9616cccadd5

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface_unittest.cc:
```diff
@@ -576,6 +576,141 @@ TEST_F(CommandLineInterfaceTest, GeneratorAndPlugin) {
+  ExpectFileContentContainsSubstring("wrapper_invoked.txt", "wrapped");
+}
+
+#ifndef _WIN32
```
Is this a fundamental limitation?  Or are these just hard to test on windows?

igor-sirotin replied:
Well scripts would need to be written differently for win/non-win, that's why I didn't bother at first.
But I agree, would be better to auto-test it on all platforms. Added in 7d5a62d6f6af163ef7972db2a17da39a70fcb440 and 3c125bd50f044bfb9f437e84b686e0497ce37308

mkruskal-google commented on src/google/protobuf/compiler/command_line_interface.cc:
```diff
@@ -2578,6 +2580,20 @@ CommandLineInterface::InterpretArgument(const std::string& name,
+          return PARSE_ARGUMENT_FAIL;
+        }
+        std::vector<std::string> tokens = absl::StrSplit(
+            value, absl::ByAnyChar(" \t"), absl::SkipWhitespace());
```
Should we error early if we detect quotes?  That would suggest someone is trying to pass an unsupported argument with whitespace and it'll blow up with an opaque error later right?

igor-sirotin replied:
Simple addition indeed. Done: 552c700fd93a434755b18c736a74492944df60f0

igor-sirotin committed:
better help

igor-sirotin committed:
Port PluginPrefixFromSearchPath test to Windows

igor-sirotin committed:
Port PluginPrefixForwardsArguments test to Windows

igor-sirotin committed:
Reject quotes in --<lang>_prefix value at parse time

igor-sirotin commented:
@mkruskal-google sorry for the delay, I addressed your last comments. Please take a look when you have a moment

igor-sirotin commented:
Also updated PR name and description.

JasonLunn commented:
@igor-sirotin - can you take a look at the test failures we're seeing on Windows, e.g.:

```
[ RUN      ] CommandLineInterfaceTest.PluginPrefixForwardsArguments
src/google/protobuf/compiler/command_line_interface_tester.cc(131): error: Expected equality of these values:
  0
  return_code_
    Which is: 1

src/google/protobuf/compiler/command_line_interface_tester.cc(139): error: Value of: line
Expected: has substring "warning:"
  Actual: "'\"baz\"' is not recognized as an internal or external command,\r"

src/google/protobuf/compiler/command_line_interface_tester.cc(139): error: Value of: line
Expected: has substring "warning:"
  Actual: "operable program or batch file.\r"

src/google/protobuf/compiler/command_line_interface_tester.cc(139): error: Value of: line
Expected: has substring "warning:"
  Actual: "--plug_out: prefix-gen-plug: Plugin failed with status code 9009."

WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
F0000 00:00:1778868726.285409    6508 mock_code_generator.cc:124] Check failed: File::GetContents( absl::StrCat(output_directory, "/", GetOutputFileName(name, file)), &content, true) is OK (NOT_FOUND: Could not open file) 
*** Check failure stack trace: ***
    @   00007FF681E2561D  (unknown)
```

igor-sirotin committed:
Fix PluginPrefixForwardsArguments wrapper on Windows

Batch %1/%2 parameter tokenization splits on '=' (and ',' ';'), so
"--bar=baz" became two parameters and the wrapper tried to exec
"baz" (exit 9009). Record the raw argument tail via %* and take the
last token as the plugin path instead.

igor-sirotin committed:
Comment bat/sh wrappers used in tests

igor-sirotin commented:
> can you take a look at the test failures we're seeing on Windows, e.g.:

@JasonLunn done! Hope it works now, unfortunately I don't have a Windows machine to verify it.

igor-sirotin commented:
@mkruskal-google @JasonLunn is it good to go now? 
I see there're some issues in `feedback/copybara` check, but I can't access details.
